File: src/hilbert/nodes/verifier.py
"""Verifier node for Hilbert."""

from typing import List

from hilbert.config.settings import get_settings
from hilbert.models import Finding, Paper
from hilbert.sources import cosine_similarity, compute_similarities, get_embedding_client
from hilbert.state.research import ResearchState


async def verifier_node(state: ResearchState) -> dict:
    """Verify claims against paper abstracts with embeddings."""
    settings = get_settings()
    threshold = settings.confidence_threshold

    findings = state.get("findings", [])
    papers = state.get("papers", [])

    if not findings or not papers:
        return {
            "findings": findings,
            "status": "writing",
        }

    paper_map = {p.paper_id: p for p in papers}

    try:
        emb_client = get_embedding_client()

        claim_texts = [f.claim for f in findings]
        context_texts = [f.evidence_text for f in findings]

        similarities = await compute_similarities(
            claims=claim_texts,
            contexts=context_texts,
            client=emb_client,
        )

        for i, finding in enumerate(findings):
            if i < len(similarities):
                finding.confidence = similarities[i]

            finding.is_verified = finding.confidence >= threshold

    except Exception:
        for finding in findings:
            finding.is_verified = False
            finding.confidence = 0.5

    return {
        "findings": findings,
        "status": "writing",
    }


def create_verifier_node():
    """Create verifier node function."""
    return verifier_node

File: cli/src/core/nodes/verifier.ts
import { getSettings } from "../config";
import { getEmbeddingClient } from "../sources/embeddings";
import type { ResearchState } from "../state";
import type { Finding } from "../models/finding";

export async function verifierNode(state: ResearchState): Promise<Partial<ResearchState>> {
  const settings = getSettings();
  const threshold = settings.confidenceThreshold;

  const findings = state.findings;
  const papers = state.papers;

  if (!findings.length || !papers.length) {
    return { findings, status: "writing" as const };
  }

  const paperMap = new Map(papers.map((p) => [p.paper_id, p]));

  try {
    const embClient = getEmbeddingClient();

    const claimTexts = findings.map((f) => f.claim);
    const contextTexts = findings.map((f) => f.evidence_text);

    const similarities = await embClient.computeSimilarities(claimTexts, contextTexts);

    for (let i = 0; i < findings.length; i++) {
      if (i < similarities.length) {
        findings[i].confidence = similarities[i];
      }
      findings[i].is_verified = findings[i].confidence >= threshold;
    }
  } catch {
    for (const finding of findings) {
      finding.is_verified = false;
      finding.confidence = 0.5;
    }
  }

  return {
    findings,
    status: "writing" as const,
  };
}

export function createVerifierNode() {
  return verifierNode;
}